### PR TITLE
[TOPIC-GPIO]: Introduce helper functions for irq flags

### PR DESCRIPTION
**Pull request by galak**
Add some helper functions to make things a bit more readable and less error prone.

### CODE REVIEW

**Review by pabigot — CHANGES_REQUESTED** [review]
I personally really dislike this sort of helper.  They're verbose, and you always have to look up the definition to figure out what it's really doing.   I'm not convinced the edge tests are at all correct as they're testing combinations of five different bit fields, some of which have an effect but aren't part of what's tested

If we're going to do this, let's wait until we see what tests people actually need (and have confirmed that they're correct) before creating new functions.

At a minimum: NAK without a test suite that verifies these do what they're supposed to do and that will detect breakage next time we have to change the bitfield assignment.

**Comment by galak** [comment]
> I personally really dislike this sort of helper. They're verbose, and you always have to look up the definition to figure out what it's really doing. I'm not convinced the edge tests are at all correct as they're testing combinations of five different bit fields, some of which have an effect but aren't part of what's tested
> 
> If we're going to do this, let's wait until we see what tests people actually need (and have confirmed that they're correct) before creating new functions.
> 
> At a minimum: NAK without a test suite that verifies these do what they're supposed to do and that will detect breakage next time we have to change the bitfield assignment.

Part of my reason for this was the number of coverity bugs I've seen in which people didn't test flags properly, especially flags that are defined with a 0 value, or if a particular flag was multiple bits.

**Comment by pabigot** [comment]
> Part of my reason for this was the number of coverity bugs I've seen in which people didn't test flags properly, especially flags that are defined with a 0 value, or if a particular flag was multiple bits.

Proper review by knowledgeable people would get rid of some of those.

While I see your point, I have no reason to believe the tests you've provided are implemented correctly, and some reason to believe they're not.

**Comment by carlescufi** [comment]
> > Part of my reason for this was the number of coverity bugs I've seen in which people didn't test flags properly, especially flags that are defined with a 0 value, or if a particular flag was multiple bits.
> 
> Proper review by knowledgeable people would get rid of some of those.
> 
> While I see your point, I have no reason to believe the tests you've provided are implemented correctly, and some reason to believe they're not.

Throwing in my 2 cents here perhaps instead of the functions having a macro that helps with the comparison:

```
#define GPIO_FLAG_IS_SET(flags, bit) ((flags & bit) == bit)

if ( GPIO_FLAG_IS_SET(flags, GPIO_INT_EDGE_RISING) {
   ...
}
```

**Comment by galak** [comment]
> owing in my 2 cents here perhaps instead of the functions having a macro that helps with the comparison:
> 
> ```
> #define GPIO_FLAG_IS_SET(flags, bit) ((flags & bit) == bit)
> 
> if ( GPIO_FLAG_IS_SET(flags, GPIO_INT_EDGE_RISING) {
>    ...
> }
> ```

I'm fine with this as well, I just want to avoid the errors we've seen with flags.

**Comment by pabigot** [comment]
OK, let's go there.

Whether it's an inline function or a macro doesn't seem important.  It's what the function does.

First, if #19248 gets in then `GPIO_INT_LEVELS_LOGICAL` is irrelevant at the driver implementation layer.  So the flags we care about are:

* `GPIO_INT_ENABLE` whether it's enabled or not
* `GPIO_INT_EDGE` whether it's edge- or level-triggered
* `GPIO_INT_LOW_0` and `GPIO_INT_HIGH_1` to indicate trigger conditions

So the mask value is wrong for the first three helper functions; rising and falling would both be true when flags is configured for both levels.  There should be a helper macro like:

    #define GPIO_INT_TRIGGER_Msk (GPIO_INT_EDGE | GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)

then the tests are things like:

    (flags & GPIO_INT_TRIGGER_Msk) == GPIO_INT_EDGE_RISING
    (flags & GPIO_INT_TRIGGER_Msk) == GPIO_INT_EDGE_BOTH
    (flags & GPIO_INT_TRIGGER_Msk) == GPIO_INT_LEVEL_LOW

Also adding:

    #define GPIO_INT_ENABLE_Msk GPIO_INT_ENABLE

then using:

    (flags & GPIO_INT_ENABLE_Msk) == GPIO_INT_ENABLE
    (flags & GPIO_INT_ENABLE_Msk) == GPIO_INT_DISABLE

is much more clear because you "know" that somebody might define `GPIO_INT_ENABLE` to have a zero value, but they won't define `GPIO_INT_ENABLE_Msk` to be zero.

Once you have those mask values, then the value of the wrapper function/macros becomes very small, IMO, because the code is clear.

**Comment by galak** [comment]
> OK, let's go there.
> 
> Whether it's an inline function or a macro doesn't seem important. It's what the function does.
> 
> First, if #19248 gets in then `GPIO_INT_LEVELS_LOGICAL` is irrelevant at the driver implementation layer. So the flags we care about are:
> 
> * `GPIO_INT_ENABLE` whether it's enabled or not
> * `GPIO_INT_EDGE` whether it's edge- or level-triggered
> * `GPIO_INT_LOW_0` and `GPIO_INT_HIGH_1` to indicate trigger conditions
> 
> So the mask value is wrong for the first three helper functions; rising and falling would both be true when flags is configured for both levels. There should be a helper macro like:
> 
> ```
> #define GPIO_INT_TRIGGER_Msk (GPIO_INT_EDGE | GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)
> ```
> 
> then the tests are things like:
> 
> ```
> (flags & GPIO_INT_TRIGGER_Msk) == GPIO_INT_EDGE_RISING
> (flags & GPIO_INT_TRIGGER_Msk) == GPIO_INT_EDGE_BOTH
> (flags & GPIO_INT_TRIGGER_Msk) == GPIO_INT_LEVEL_LOW
> ```
> 
> Also adding:
> 
> ```
> #define GPIO_INT_ENABLE_Msk GPIO_INT_ENABLE
> ```
> 
> then using:
> 
> ```
> (flags & GPIO_INT_ENABLE_Msk) == GPIO_INT_ENABLE
> (flags & GPIO_INT_ENABLE_Msk) == GPIO_INT_DISABLE
> ```
> 
> is much more clear because you "know" that somebody might define `GPIO_INT_ENABLE` to have a zero value, but they won't define `GPIO_INT_ENABLE_Msk` to be zero.
> 
> Once you have those mask values, then the value of the wrapper function/macros becomes very small, IMO, because the code is clear.

I find it a little weird to see:

`(flags & GPIO_INT_ENABLE_Msk) == GPIO_INT_DISABLE`

since the 'Msk' name doesn't match the thing I'm testing.  Just my 2 cents.

**Comment by pabigot** [comment]
> I find it a little weird to see:
> 
> `(flags & GPIO_INT_ENABLE_Msk) == GPIO_INT_DISABLE`
> 
> since the 'Msk' name doesn't match the thing I'm testing. Just my 2 cents.

You're not wrong.

What I did was modeled on the CMSIS standard for flags, where you'd have:
```
#define GPIO_INT_ENABLE_Pos 12
#define GPIO_INT_ENABLE_Msk (1U << GPIO_INT_ENABLE_Pos)
#define GPIO_INT_ENABLE_Disabled 0U
#define GPIO_INT_ENABLE_Enabled 1U
```

then you'd test with:

    (flags & GPIO_INT_ENABLE_Msk) == (GPIO_INT_ENABLE_Disabled << GPIO_INT_ENABLE_Pos)

Which is uniform and sensible (though a little annoying to have to write the value shift).

But that's not what we have to work with and I'm not going to argue for changing all the flag names at this point.

File: drivers/gpio/gpio_utils.h [hunk 1]
```diff
@@ -64,4 +65,39 @@ static inline void gpio_fire_callbacks(sys_slist_t *list,
 	}
 }
 
+static inline bool gpio_irq_flags_edge_both(int flags)
```
Comment (pabigot) [inline drivers/gpio/gpio_utils.h]: How about `gpio_flags_int_edge_both`?

First, the flags are `GPIO_INT_foo` not `GPIO_IRQ_foo`, so using `irq` is confusing.

Second, we might want similar functionality for other parts of the flags, e.g. `gpio_flags_pull_up()`.  It'd be nice if functionally-related operations shared a common prefix.

**Comment by galak** [comment]
Just add a reference to a slew of fixes (might have a few mistakes) to show the kinda issues we've had with `#define FLAG 0`

61fb6b468d016ef77fa9022fd75514ec88bbedf4
1af5fa174b4b8b24e299606bb6ecc2ea16989a8c
69d4d32acf4c38e915f931163554285af04c26d6
dc45def9d0a413a9381a4f1e1dcda515eec30035
3d48ce801108b9802a143f388cdb5b0546c7b968
0b18f6dd39199f75364a8abf1ee16539da9219c7
be55f6644e8579aa722fbed2202830608e786e44
4b288999b2f68d1030d3e6aaf92e7ab4443f4fc8
b52080749fb6a88038f88a53b123c0dd67012795
6b04e0669a4dfbc428e7ba4201f61985001a9d8a
cd63c74bbdfdd1ef654b10f017c0811775211b40
6c393fa3931f0823d3a4bba9518f9997e1cf1574
23a0f6c918dcbc90f6f2dbdd256f5c78ff6261a0
97c5020079d5eb63a357d5124669781f130972c5

File: drivers/gpio/gpio_utils.h (hunk 1 above)
Comment (galak) [reply drivers/gpio/gpio_utils.h]: Done

**Comment by pabigot** [comment]
I'm not disagreeing with your goal.  Some of those are probably from me, or in response to issues I opened because I'm very sensitive to this problem.

But if this is going to be solved with helper functions/macros they must be correct, and the implementation in this PR is not correct, and it has no tests to ensure it would remain correct.

**Comment by carlescufi** [comment]
I am not going to state any particular opinion for or against this proposal because I can't really see whether it will or will not be better at avoiding bugs like the ones @galak listed in the future. Everybody sees code differently and what is obvious to one person looking at bitmasks is a puzzle for others unless they have a function. Perhaps providing both options and not insisting in using one or the other is the better way to go, but even then I am on the fence.

**Comment by galak** [comment]
> I'm not disagreeing with your goal. Some of those are probably from me, or in response to issues I opened because I'm very sensitive to this problem.
> 
> But if this is going to be solved with helper functions/macros they must be correct, and the implementation in this PR is not correct, and it has no tests to ensure it would remain correct.

I pushed a test for the functions.  If there is something that is wrong with them please comment on what it is.

**Comment by galak** [comment]
> I am not going to state any particular opinion for or against this proposal because I can't really see whether it will or will not be better at avoiding bugs like the ones @galak listed in the future. Everybody sees code differently and what is obvious to one person looking at bitmasks is a puzzle for others unless they have a function. Perhaps providing both options and not insisting in using one or the other is the better way to go, but even then I am on the fence.

I disagree.  I think we've seen a common pattern bug with a flag that gets defined as 0, and developers not handling it properly repeatedly.  I'm open to any solution to the repeated problem and I'm proposing one that I think would reduce the errors.

**Comment by pabigot** [comment]
> I pushed a test for the functions. If there is something that is wrong with them please comment on what it is.

Never seen that kind of unit test in Zephyr before.

What's wrong is that these undocumented functions do not do what **this** user would expect.

When I'm figuring how to configure edge interrupts, I'm looking to see which of the three cases apply: rising edge, falling edge, or both.  In other words, did the user provide `GPIO_INT_EDGE_RISING`, `GPIO_INT_EDGE_FALLING`, or `GPIO_INT_EDGE_BOTH`.  Because those are the three distinct values people are most likely to provide.

The test reveals that you have a different interpretation which is based on handling the individual flag bits, rather than the overall configuration.  I would have coded the test as:

```
	flags = GPIO_INT_EDGE_BOTH;
	zassert_true(gpio_flags_int_edge_both(flags), NULL);
	zassert_false(gpio_flags_int_edge_rising(flags), NULL); // both != rising
	zassert_false(gpio_flags_int_edge_falling(flags), NULL); // both != falling
```

which would have failed.

It's exactly this sort of thing that I hate about these helper functions and is why I always have to look at them to see if they're doing what I expect them to do.  Too often they don't.

In any case, you seem to really want these, and you have provided a test that confirms they do what you intend.  I don't think they're a good solution to the problem, but if you get other people to approve I'll remove my block.

**Comment by galak** [comment]
> Never seen that kind of unit test in Zephyr before.
> 
Wasn't sure we had anything like it.

> What's wrong is that these undocumented functions do not do what **this** user would expect.
> 
> When I'm figuring how to configure edge interrupts, I'm looking to see which of the three cases apply: rising edge, falling edge, or both. In other words, did the user provide `GPIO_INT_EDGE_RISING`, `GPIO_INT_EDGE_FALLING`, or `GPIO_INT_EDGE_BOTH`. Because those are the three distinct values people are most likely to provide.
> 
> The test reveals that you have a different interpretation which is based on handling the individual flag bits, rather than the overall configuration. I would have coded the test as:
> 
> ```
> 	flags = GPIO_INT_EDGE_BOTH;
> 	zassert_true(gpio_flags_int_edge_both(flags), NULL);
> 	zassert_false(gpio_flags_int_edge_rising(flags), NULL); // both != rising
> 	zassert_false(gpio_flags_int_edge_falling(flags), NULL); // both != falling
> ```
> 
> which would have failed.
> 

Fixed up to the behavior you suggest which makes sense to me.

**Comment by carlescufi** [comment]
> > I am not going to state any particular opinion for or against this proposal because I can't really see whether it will or will not be better at avoiding bugs like the ones @galak listed in the future. Everybody sees code differently and what is obvious to one person looking at bitmasks is a puzzle for others unless they have a function. Perhaps providing both options and not insisting in using one or the other is the better way to go, but even then I am on the fence.
> 
> I disagree. I think we've seen a common pattern bug with a flag that gets defined as 0, and developers not handling it properly repeatedly. I'm open to any solution to the repeated problem and I'm proposing one that I think would reduce the errors.

But we can´t know how many issues we would have found if we had had those static inlines. It's a fair point, all I am stating is that I do not have a strong opinion in favor or against and I am not fully convinced this will improve the final bug count.

**Pull request closed without merging**